host_applications/linux/apps/hello_pi/hello_mmal_encode/mmal_encode.c

```diff
@@ -0,0 +1,268 @@
+    exit(1);
+  }
+
+  portIn->buffer_size = portIn->buffer_size_recommended;
+  portIn->buffer_num = portIn->buffer_num_recommended;
+
+  if (mmal_wrapper_port_enable(portIn, MMAL_WRAPPER_FLAG_PAYLOAD_ALLOCATE)
+      != MMAL_SUCCESS) {
+    fprintf(stderr, "Failed to enable input port\n");
+    exit(1);
+  }
+
+  // Create a test input image consisting of 8 vertical bars of
+  // white | black | red | green | blue | cyan | magenta | yellow
+
+  inBufSize = portIn->format->es->video.width * portIn->format->es->video.height * BYTESPP;
+  inBuf = malloc(inBufSize);
+  inPix = (uint32_t*)inBuf;
+  for (i=0; i<inBufSize/BYTESPP; ++i) {
+    switch ((i % portIn->format->es->video.width) / (WIDTH / 8)) {
+    case 0: *inPix = RED | GREEN | BLUE; break;
+    case 1: *inPix = 0; break;
+    case 2: *inPix = RED; break;
+    case 3: *inPix = GREEN; break;
+    case 4: *inPix = BLUE; break;
+    case 5: *inPix = GREEN | BLUE; break;
+    case 6: *inPix = RED | BLUE; break;
+    case 7: *inPix = RED | GREEN; break;
+    }
+    *inPix |= ALPHA; // Full alpha
+    ++inPix;
+  }
+  
+  fprintf(stdout, "- input %4.4s %ux%u\n",
+          (char*)&portIn->format->encoding,
+          portIn->format->es->video.width, portIn->format->es->video.height);
+
+  // Configure output
+  
+  portOut = mmal->output[0];
+
+  if (portOut->is_enabled) {
+    if (mmal_wrapper_port_disable(portOut) != MMAL_SUCCESS) {
+      fprintf(stderr, "Failed to disable output port\n");
+      exit(1);
+    }
+  }
+
+  portOut->format->encoding = encoding;
+  if (mmal_port_format_commit(portOut) != MMAL_SUCCESS) {
+    fprintf(stderr, "Failed to commit output port format\n");
+    exit(1);
+  }
+
+  mmal_port_parameter_set_uint32(portOut, MMAL_PARAMETER_JPEG_Q_FACTOR, 100);
+  
+  portOut->buffer_size = portOut->buffer_size_recommended;
+  portOut->buffer_num = portOut->buffer_num_recommended;
+
+  if (mmal_wrapper_port_enable(portOut, 0) != MMAL_SUCCESS) {
+    fprintf(stderr, "Failed to enable output port\n");
+    exit(1);
+  }
+
+  fprintf(stdout, "- output %4.4s\n", (char*)&encoding);
+  
+  // Perform the encoding
+
+  outBuf = malloc(portOut->buffer_size);
+  outFile = open(filename, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+  if (outFile < 0) {
+    fprintf(stderr, "Failed to open file: %s\n", filename);
+    exit(1);
+  }
+  
+  while (!eos) {
+
+    // Send output buffers to be filled with encoded image.
+    while (mmal_wrapper_buffer_get_empty(portOut, &out, 0) == MMAL_SUCCESS) {
+      out->data = outBuf;
+      out->alloc_size = portOut->buffer_size;
+      if (mmal_port_send_buffer(portOut, out) != MMAL_SUCCESS) {
+        fprintf(stderr, "Failed to send buffer\n");
+        break;
+      }
+    }
+
+    // Send image to be encoded.
+    if (!sent && mmal_wrapper_buffer_get_empty(portIn, &in, 0) == MMAL_SUCCESS) {
+      fprintf(stdout, "- sending %u bytes to encoder\n", inBufSize);
+      memcpy(in->data, inBuf, inBufSize);
```

Comment:
I could construct the test image directly in the in->data buffer.
I'm also assuming the size of this buffer is always going to be >= inBuffSize (which it does always seem to be), but not sure if this is guaranteed?